Index: src/game/g_active.c
===================================================================
--- src/game/g_active.c	(revision 67)
+++ src/game/g_active.c	(working copy)
@@ -570,6 +570,7 @@
 */
 void ClientTimerActions( gentity_t *ent, int msec )
 {
+  char message[ MAX_STRING_CHARS ];
   gclient_t *client;
   usercmd_t *ucmd;
   int       aForward, aRight;
@@ -845,6 +846,24 @@
   {
     client->time1000 -= 1000;
 
+    //Don't leave your base before !INVASION START :p
+    if( ent->client->ps.stats[ STAT_HEALTH ] > 0 && ent->client->ps.stats[ STAT_PTEAM ] == PTE_HUMANS &&
+      !(ent->r.svFlags & SVF_BOT) && (level.numHumanSpawns != 0) && g_invasion.integer )
+    {
+      if( !G_BuildableRange( ent->client->ps.origin, 900, BA_H_REACTOR ) )
+      {
+        if( !(client->ps.stats[ STAT_STATE ] & SS_GRABBED) )
+        {
+           trap_SendServerCommand( -1, "cp \"^1Don't leave your base before invasion starts\n\"" );
+           Com_sprintf( message, sizeof( message ), "%s^7 tried to leave base before invasion starts\n", client->pers.netname );
+           G_AdminsPrintf(message);
+        }
+        client->grabExpiryTime = level.time + 1500;	//grabbed until !INVASION START is done
+        client->ps.stats[ STAT_STATE ] = SS_GRABBED;
+        G_Damage( ent, NULL, NULL, NULL, NULL, 2, DAMAGE_NO_PROTECTION, MOD_UNKNOWN );
+      }
+    }
+
     //client is poison clouded
     if( client->ps.stats[ STAT_STATE ] & SS_POISONCLOUDED )
       G_Damage( ent, client->lastPoisonCloudedClient, client->lastPoisonCloudedClient, NULL, NULL,
Index: src/game/g_admin.c
===================================================================
--- src/game/g_admin.c	(revision 67)
+++ src/game/g_admin.c	(working copy)
@@ -158,7 +158,18 @@
       "display the contents of server info files",
       "(^5subject^7)"
     },
-    
+
+    {"invasion", G_admin_invasion, "invasion",
+      "Load or Start the scenario ^2I^7nvasion",
+      "[^3nb_bot^7] [^3name_bot^7] (^5layout^7)"
+      "\n ^3Syntax:^7 !invasion [^3start^7] => Destruction of the telenodes"
+    },
+
+    {"listinvasion", G_admin_listinvasion, "listinvasion",
+      "display a list of all available layouts ^2I^7nvasion",
+      "(^5mapname^7)"
+    },
+
     {"invisible", G_admin_invisible, "invisible",
       "hides a player so they cannot be seen in playerlists",
       ""
@@ -3274,6 +3285,12 @@
     return qfalse;
   }
 
+  if (g_invasion.integer && team[ 0 ] != 's')
+  {
+     ADMP( va( "^3!putteam: ^7Invasion is running, you can only putteam someone in spectators team\n" ) );
+    return qfalse;
+  }
+
   switch( team[ 0 ] )
   {
   case 'a':
@@ -3744,7 +3761,7 @@
   }
 
   G_SayArgv( skiparg + 1, map, sizeof( map ) );
-
+  Q_strlwr(map);
   if( !trap_FS_FOpenFile( va( "maps/%s.bsp", map ), NULL, FS_READ ) )
   {
     ADMP( va( "^3!map: ^7invalid map name '%s'\n", map ) );
@@ -3754,6 +3771,8 @@
   if( G_SayArgc( ) > 2 + skiparg )
   {
     G_SayArgv( skiparg + 2, layout, sizeof( layout ) );
+    Q_strlwr(layout);
+    Q_strlwr(layout);
     if( !Q_stricmp( layout, "*BUILTIN*" ) ||
       trap_FS_FOpenFile( va( "layouts/%s/%s.dat", map, layout ),
         NULL, FS_READ ) > 0 )
@@ -4012,7 +4031,7 @@
   }
 
   G_SayArgv( skiparg + 1, layout, sizeof( layout ) );
-
+  Q_strlwr(layout);
   trap_SendConsoleCommand( EXEC_APPEND, va( "layoutsave %s", layout ) );
   AP( va( "print \"^3!layoutsave: ^7layout saved as '%s' by %s\n\"", layout,
           ( ent ) ? G_admin_adminPrintName( ent ) : "console" ) );
@@ -4536,6 +4555,7 @@
     G_SayArgv( 1 +skiparg, map, sizeof( map ) );
   else
     trap_Cvar_VariableStringBuffer( "mapname", map, sizeof( map ) );
+  Q_strlwr(map);
   
   count = G_LayoutList( map, list, sizeof( list ) );
   ADMBP_begin( );
@@ -5816,8 +5836,9 @@
     char map[ MAX_QPATH ];
 
     trap_Cvar_VariableStringBuffer( "mapname", map, sizeof( map ) );
+    Q_strlwr(map);
     G_SayArgv( skiparg + 1, layout, sizeof( layout ) );
-
+    Q_strlwr(layout);
     if( Q_stricmp( layout, "keepteams" ) && Q_stricmp( layout, "keepteamslock" ) && Q_stricmp( layout, "switchteams" ) && Q_stricmp( layout, "switchteamslock" ) )
     {
       if( !Q_stricmp( layout, "*BUILTIN*" ) ||
@@ -6173,6 +6194,11 @@
       ADMP( "^3!unlock: ^7Alien team is not currently locked\n" );
       return qfalse;
     }
+    else if (g_invasion.integer)
+    {
+      ADMP( "^3!unlock: ^7Invasion is running, you can't unlock alien team.\n" );
+      return qfalse;
+    }
     else
       level.alienTeamLocked = qfalse;
   }
@@ -6182,6 +6208,11 @@
       ADMP( "^3!unlock: ^7Human team is not currently locked\n" );
       return qfalse;
     }
+    else if (g_invasion.integer)
+    {
+      ADMP( "^3!unlock: ^7Invasion is running, you can't unlock human team.\n" );
+      return qfalse;
+    }
     else
       level.humanTeamLocked = qfalse;
   }
@@ -6688,6 +6719,11 @@
     ADMP( "^3!revert: ^7no build log found\n" );
     return qfalse;
   }
+  if( g_invasion.integer )
+  {
+    ADMP( "^3!revert: ^7Invasion is running, you can't use !revert .\n" );
+    return qfalse;
+  }
   if( G_SayArgc( ) < 2 + skiparg )
   {
     ADMP( "^3!revert: ^7usage: !revert (^5xnum^7) (^5#ID^7) (^5-name|num^7) (^5a|h^7)\n" );
@@ -7235,6 +7271,155 @@
   return qtrue;
 }
 
+qboolean G_admin_invasion( gentity_t *ent, int skiparg )
+{
+  int nb_bot_int = 0;
+  char arg1[ MAX_STRING_TOKENS ] = {""};
+  char layout[ MAX_STRING_TOKENS ] = {""};
+  char name_bot[ MAX_NAME_LENGTH ] = {""};
+  char map[ MAX_QPATH ] = {""};
+  
+  G_SayArgv( 1 + skiparg, arg1, sizeof( arg1 ) );
+  if( (G_SayArgc() < 3 + skiparg) && (Q_stricmp( arg1, "start" )) )
+  {
+    ADMP( "^3!invasion: ^7!invasion [^3nb_bot^7] [^3name_bot^7] (^5layout^7) \n" );
+    ADMP( "^3!invasion: ^7!invasion [^3start^7] => Destruction of the telenodes\n" );
+    return qfalse;
+  }
+  else
+  {
+     if( ( G_SayArgc() == 2 + skiparg ) && !Q_stricmp( arg1, "start" ) )
+     {
+       if( (g_invasion.integer) && (level.numHumanSpawns != 0) )
+       {
+         level.humanTeamLocked = qtrue;
+         trap_SendConsoleCommand( EXEC_APPEND, "alienWin\n" );
+         trap_SendServerCommand( -1, "cp \"^4 Start !!!\n\"" );
+         return qtrue;
+       }
+       else
+       {
+         ADMP( "^7Scenario ^2I^7nvasion can't start now\n" );
+         return qfalse;
+       }
+     }
+     else if( G_SayArgc() >= 3 + skiparg )
+     {
+       nb_bot_int = atoi(arg1);
+       G_SayArgv( 2 + skiparg, name_bot, sizeof( name_bot ) );
+       if( G_SayArgc() == 4 + skiparg )
+         G_SayArgv( 3 + skiparg, layout, sizeof( layout ) );
+     }
+  }
+  if( !strlen( layout ) )
+    Q_strncpyz(layout, g_invasionLayout.string , sizeof(layout));
+  Q_strlwr(layout);
+  trap_Cvar_VariableStringBuffer( "mapname", map, sizeof( map ) );
+  Q_strlwr(map);
+
+  if( !trap_FS_FOpenFile( va( "layouts/%s/%s.dat", map, layout ), NULL, FS_READ ) ||
+	 Q_strncmp( layout, Q_strlwr(g_invasionLayout.string), strlen(g_invasionLayout.string) ) )
+  {
+    ADMP( "^7Scenario ^2I^7nvasion^1 could not be done on this map and/or on this layout\n" );
+    return qfalse;
+  }
+  if( (nb_bot_int < g_invasionBotMin.integer) || ( nb_bot_int > g_invasionBotMax.integer ) )
+  {
+    ADMP( va( "The number of bots must be between %i and %i\n", g_invasionBotMin.integer, g_invasionBotMax.integer ) );
+    return qfalse;
+  }
+  trap_Cvar_Set( "g_invasionBotNumber" , va("%i", nb_bot_int) );
+
+  // Prefix of the invasion's bots
+  trap_Cvar_Set( "g_invasionBotPrefix", name_bot );
+
+  // Current Layout
+  trap_Cvar_Set( "g_invasionCurrentLayout" , layout );
+
+  // Time in seconds before start invasion  (between 20 and 60)
+  if( g_invasionStartTime.integer < 20 )
+    trap_Cvar_Set( "g_invasionStartTime" , va("%i", 20) );
+  else if( g_invasionStartTime.integer > 60 )
+    trap_Cvar_Set( "g_invasionStartTime" , va("%i", 60) );
+
+  ADMP( va("^7Scenario ^2I^7nvasion : ^3nb_bot^2=^7%i  ^3name_bot^2=^7%s  ^3layout^2=^7%s\n", nb_bot_int, name_bot, layout) );
+  trap_Cvar_Set( "g_layouts", layout );
+  trap_Cvar_Set( "g_lockTeamsAtStart", "1" );
+  trap_Cvar_Set( "g_invasion", "1" );
+  trap_SendConsoleCommand( EXEC_APPEND, "map_restart 0\n" );
+
+  return qtrue;
+}
+
+#define MAX_LISTINVASION 64
+
+qboolean G_admin_listinvasion( gentity_t *ent, int skiparg )
+{
+  char fileList[ 4096 ] = {""};
+  char *fileSort[ MAX_LISTINVASION ];
+  char map[ MAX_QPATH ] = {""};
+  int numFiles;
+  int i;
+  int fileLen = 0;
+  int  count = 0;
+  char *filePtr;
+  int rows;
+
+  if( G_SayArgc( ) > 1 + skiparg )
+  {
+    G_SayArgv( skiparg + 1, map, sizeof( map ) );
+    Q_strlwr(map);
+  }
+  else
+  {
+    trap_Cvar_VariableStringBuffer( "mapname", map, sizeof( map ) );
+    Q_strlwr(map);
+    if( !map[ 0 ] )
+    {
+      G_Printf( "^3!listinvasion: ^7no map is loaded\n" );
+      return qfalse;
+    }
+  }
+
+  numFiles = trap_FS_GetFileList( va("layouts/%s/", map), ".dat",
+    fileList, sizeof( fileList ) );
+  filePtr = fileList;
+  for( i = 0; i < numFiles && count < MAX_LISTINVASION; i++, filePtr += fileLen + 1 )
+  {
+    fileLen = strlen( filePtr );
+    if (fileLen < 5)
+      continue;
+
+    filePtr[ fileLen - 4 ] = '\0';
+
+    if( Q_strncmp( filePtr, Q_strlwr(g_invasionLayout.string), strlen(g_invasionLayout.string) ) )
+      continue;
+
+    fileSort[ count ] = filePtr;
+    count++;
+  }
+
+  qsort(fileSort, count, sizeof(fileSort[ 0 ]), SortMaps);
+
+  rows = count / 3;
+  if ( rows * 3 < count ) rows++;
+
+  ADMBP_begin();
+  for( i = 0; i < rows; i++ )
+  {
+    ADMBP( va( "^7%20s %20s %20s\n",
+      fileSort[ i ],
+      ( rows + i < count ) ? fileSort[ rows + i ] : "",
+      ( rows * 2 + i < count ) ? fileSort[ rows * 2 + i ] : "" ) );
+  }
+
+  ADMBP( va( "^3!listinvasion: ^7found %d layouts ^2I^7nvasion in '%s^7' map.\n", count, map ) );
+
+  ADMBP_end();
+
+  return qtrue;
+}
+
 qboolean G_admin_invisible( gentity_t *ent, int skiparg )
 {
   if( !ent )
Index: src/game/g_admin.h
===================================================================
--- src/game/g_admin.h	(revision 67)
+++ src/game/g_admin.h	(working copy)
@@ -252,6 +252,8 @@
 qboolean G_admin_warn( gentity_t *ent, int skiparg );
 qboolean G_admin_designate( gentity_t *ent, int skiparg );
 qboolean G_admin_cp( gentity_t *ent, int skiparg );
+qboolean G_admin_invasion( gentity_t *ent, int skiparg );
+qboolean G_admin_listinvasion( gentity_t *ent, int skiparg );
 
 qboolean G_admin_slap( gentity_t *ent, int skiparg );
 qboolean G_admin_drop( gentity_t *ent, int skiparg );
Index: src/game/g_buildable.c
===================================================================
--- src/game/g_buildable.c	(revision 67)
+++ src/game/g_buildable.c	(working copy)
@@ -1962,7 +1962,8 @@
 
         if( player->client && player->client->ps.stats[ STAT_PTEAM ] == PTE_HUMANS )
         {
-          if( player->health < player->client->ps.stats[ STAT_MAX_HEALTH ] &&
+          if( ( player->health < player->client->ps.stats[ STAT_MAX_HEALTH ] ||
+            ( player->client->ps.stats[ STAT_STAMINA ] < MAX_STAMINA && g_invasion.integer ) ) &&
               player->client->ps.pm_type != PM_DEAD )
           {
             self->enemy = player;
@@ -1998,6 +1999,19 @@
 
       self->enemy->health++;
 
+      if (g_invasion.integer) {
+        if (self->enemy->health < self->enemy->client->ps.stats[ STAT_MAX_HEALTH ]) 
+          self->enemy->health+=4;
+        if (self->enemy->client->ps.stats[ STAT_STAMINA ] < MAX_STAMINA)
+          self->enemy->client->ps.stats[ STAT_STAMINA ] += 100;
+      }
+
+      //if hp > MAX or stamina > MAX
+      if (self->enemy->health >= self->enemy->client->ps.stats[ STAT_MAX_HEALTH ]) 
+        self->enemy->health = self->enemy->client->ps.stats[ STAT_MAX_HEALTH ];
+      if (self->enemy->client->ps.stats[ STAT_STAMINA ] > MAX_STAMINA)
+        self->enemy->client->ps.stats[ STAT_STAMINA ] = MAX_STAMINA;
+
       //if they're completely healed, give them a medkit
       if( self->enemy->health >= self->enemy->client->ps.stats[ STAT_MAX_HEALTH ] &&
           !BG_InventoryContainsUpgrade( UP_MEDKIT, self->enemy->client->ps.stats ) )
Index: src/game/g_client.c
===================================================================
--- src/game/g_client.c	(revision 67)
+++ src/game/g_client.c	(working copy)
@@ -1731,8 +1731,10 @@
 
       if( spawnPoint->biteam == PTE_ALIENS )
         spawnPoint->clientSpawnTime = ALIEN_SPAWN_REPEAT_TIME;
-      else if( spawnPoint->biteam == PTE_HUMANS )
+      else if( spawnPoint->biteam == PTE_HUMANS && !g_invasion.integer)
         spawnPoint->clientSpawnTime = HUMAN_SPAWN_REPEAT_TIME;
+      else if( spawnPoint->biteam == PTE_HUMANS && g_invasion.integer)
+        spawnPoint->clientSpawnTime = 2000;
     }
   }
   client->pers.teamState.state = TEAM_ACTIVE;
Index: src/game/g_cmds.c
===================================================================
--- src/game/g_cmds.c	(revision 67)
+++ src/game/g_cmds.c	(working copy)
@@ -824,12 +824,18 @@
       return;
     }
 
-    if( level.alienTeamLocked && !force )
+    if( level.alienTeamLocked && !force && !g_invasion.integer)
     {
       trap_SendServerCommand( ent-g_entities,
         va( "print \"Alien team has been ^1LOCKED\n\"" ) );
       return; 
     }
+    else if( level.alienTeamLocked && g_invasion.integer)
+    {
+      trap_SendServerCommand( ent-g_entities,
+        "print \"Alien team has been ^1LOCKED^7 because ^2INVASION^7 is running ! Read !info invasion and wait next game for join.\n\"" );
+      return; 
+    }
     else if( level.humanTeamLocked )
     {
       // if only one team has been locked, let people join the other
@@ -854,12 +860,18 @@
       return;
     }
 
-    if( level.humanTeamLocked && !force )
+    if( level.humanTeamLocked && !force && !g_invasion.integer)
     {
       trap_SendServerCommand( ent-g_entities,
         va( "print \"Human team has been ^1LOCKED\n\"" ) );
       return; 
     }
+    else if( level.humanTeamLocked && g_invasion.integer)
+    {
+      trap_SendServerCommand( ent-g_entities,
+        "print \"Human team has been ^1LOCKED^7 because ^2INVASION^7 is running ! Read !info invasion and wait next game for join.\n\"" );
+      return; 
+    }
     else if( level.alienTeamLocked )
     {
       // if only one team has been locked, let people join the other
@@ -1821,6 +1833,83 @@
         sizeof( level.voteDisplayString ), "Set the next map to '%s^7'", arg2 );
     level.votePassThreshold = g_mapVotesPercent.integer;
   }
+  else if( !Q_stricmp( arg1, "invasion" ) )
+  {
+    char map[ MAX_QPATH ], layout[ MAX_STRING_TOKENS ], arg3[ MAX_STRING_TOKENS ];
+    int nb_bot_int = 0;
+	 int i;
+    qboolean numeric = qtrue;
+
+    trap_Argv( 3, arg3, sizeof( arg3 ) );
+
+    if( g_mapvoteMaxTime.integer 
+      && (( level.time - level.startTime ) >= g_mapvoteMaxTime.integer * 1000 )
+      && !G_admin_permission( ent, ADMF_NO_VOTE_LIMIT ) 
+      && (level.numPlayingClients > 0 && level.numConnectedClients>1) )
+    {
+      trap_SendServerCommand( ent-g_entities, va(
+        "print \"^1You cannot call for the ^7Scenario ^2I^7nvasion ^1after %d seconds\n\"",
+        g_mapvoteMaxTime.integer ) );
+      return;
+    }
+    if( trap_Argc() == 2)
+    {
+      nb_bot_int = 8;
+      Q_strncpyz(layout, g_invasionLayout.string , sizeof(layout));
+      trap_SendServerCommand( ent - g_entities, "print \"callvote: "
+        "^7Use /callvote invasion (^5nb_bot^7) (^5layout^7) \n\"" );
+    }
+    else if( trap_Argc() == 3 )
+    {
+      for( i = 0; i < sizeof( arg2 ) && arg2[ i ] ; i++ )
+      {
+        if( !Q_isdigit(arg2[ i ]) )
+        {
+          numeric = qfalse;
+          break;
+        }
+      }
+      if( !numeric )
+      {
+        Q_strncpyz(layout, arg2, sizeof(layout));
+        nb_bot_int = 8;
+      }
+      else
+      {
+        nb_bot_int = atoi(arg2);
+        Q_strncpyz(layout, g_invasionLayout.string , sizeof(layout));
+      }
+    }
+    else if( trap_Argc() == 4 )
+    {
+      Q_strncpyz(layout, arg3, sizeof(layout));
+      nb_bot_int = atoi(arg2);
+    }
+
+	Q_strlwr(layout);
+    trap_Cvar_VariableStringBuffer( "mapname", map, sizeof( map ) );
+    Q_strlwr(map);
+
+    if( !trap_FS_FOpenFile( va( "layouts/%s/%s.dat", map, layout ), NULL, FS_READ ) ||
+	   Q_strncmp( layout, Q_strlwr(g_invasionLayout.string), strlen(g_invasionLayout.string) ) )
+    {
+      trap_SendServerCommand( ent - g_entities, "print \"callvote: "
+        "^7Scenario ^2I^7nvasion^7 could not be done on this map and/or on this layout\n\"" );
+      return;
+    }
+
+    if( (nb_bot_int < g_invasionBotMin.integer) || ( nb_bot_int > g_invasionBotMax.integer ) )
+    {
+      trap_SendServerCommand( ent - g_entities, va( "print \"callvote: "
+        "The number of bots must be between %i and %i\n", g_invasionBotMin.integer, g_invasionBotMax.integer ) );
+      return;
+    }
+
+    Com_sprintf( level.voteString, sizeof( level.voteString ), va("!invasion %i ^2I^7nvasionBot_^2# %s\n", nb_bot_int, layout) );
+    Com_sprintf( level.voteDisplayString,
+        sizeof( level.voteDisplayString ), "Start the ^7Scenario ^2I^7nvasion with ^2%i^7 Bots on layout ^2%s^7", nb_bot_int, layout );
+    level.votePassThreshold = g_mapVotesPercent.integer;
+  }
   else if( !Q_stricmp( arg1, "draw" ) )
   {
     Com_sprintf( level.voteString, sizeof( level.voteString ), "evacuation" );
@@ -1993,7 +2082,7 @@
     {
       trap_SendServerCommand( ent-g_entities, "print \"Invalid vote string\n\"" );
       trap_SendServerCommand( ent-g_entities, "print \"Valid vote commands are: "
-        "map, map_restart, draw, extend, nextmap, kick, spec, mute, unmute, poll, and sudden_death\n" );
+        "map, map_restart, invasion, draw, extend, nextmap, kick, spec, mute, unmute, poll, and sudden_death\n" );
       if( customVoteKeys[ 0 ] != '\0' )
         trap_SendServerCommand( ent-g_entities,
           va( "print \"Additional custom vote commands: %s\n\"", customVoteKeys ) );
@@ -2730,8 +2819,8 @@
     else if( ent->client->pers.teamSelection == PTE_HUMANS )
     {
       //set the item to spawn with
-      if( !Q_stricmp( s, BG_FindNameForWeapon( WP_MACHINEGUN ) ) &&
-          BG_WeaponIsAllowed( WP_MACHINEGUN ) )
+      if( ( !Q_stricmp( s, BG_FindNameForWeapon( WP_MACHINEGUN ) ) &&
+          BG_WeaponIsAllowed( WP_MACHINEGUN ) ) || g_invasion.integer )
       {
         ent->client->pers.humanItemSelection = WP_MACHINEGUN;
       }
@@ -2894,7 +2983,7 @@
       ent->client->ps.stats[ STAT_PCLASS ] = PCL_HUMAN;
 
     //set the item to spawn with
-    if( !Q_stricmp( s, BG_FindNameForWeapon( WP_MACHINEGUN ) ) && BG_WeaponIsAllowed( WP_MACHINEGUN ) )
+    if( ( !Q_stricmp( s, BG_FindNameForWeapon( WP_MACHINEGUN ) ) && BG_WeaponIsAllowed( WP_MACHINEGUN ) ) || g_invasion.integer )
       ent->client->pers.humanItemSelection = WP_MACHINEGUN;
     else if( !Q_stricmp( s, BG_FindNameForWeapon( WP_HBUILD ) ) && BG_WeaponIsAllowed( WP_HBUILD ) )
       ent->client->pers.humanItemSelection = WP_HBUILD;
@@ -3422,6 +3511,13 @@
       return;
     }
 
+    //are we /allowed/ to buy this when g_invasion?
+    if ( ( g_invasion.integer && ( (weapon == WP_FLAMER) || (weapon == WP_HBUILD) || (weapon == WP_HBUILD2) ) ) )
+    {
+      trap_SendServerCommand( ent-g_entities, "print \"You can't buy this item during Scenario Invasion\n\"" );
+      return;
+    }
+
     //add to inventory
     BG_AddWeaponToInventory( weapon, ent->client->ps.stats );
     BG_FindAmmoForWeapon( weapon, &maxAmmo, &maxClips );
@@ -3500,6 +3596,13 @@
       return;
     }
 
+    //are we /allowed/ to buy this when g_invasion?
+    if( g_invasion.integer && upgrade == UP_JETPACK)
+    {
+      trap_SendServerCommand( ent-g_entities, "print \"You can't buy this item during Scenario Invasion\n\"" );
+      return;
+    }
+
     if( upgrade == UP_AMMO )
       G_GiveClientMaxAmmo( ent, buyingEnergyAmmo );
     else
Index: src/game/g_combat.c
===================================================================
--- src/game/g_combat.c	(revision 67)
+++ src/game/g_combat.c	(working copy)
@@ -1319,7 +1319,7 @@
          else if(g_friendlyFireMovementAttacks.value > 0 && g_friendlyFireMovementAttacks.value < 1)
            damage =(int)(0.5 + g_friendlyFireMovementAttacks.value * (float) damage);
       }
-      if( g_friendlyBuildableFire.value <= 0 )
+      if( g_friendlyBuildableFire.value <= 0 || g_invasion.integer )
       {
         return;
       }
Index: src/game/g_local.h
===================================================================
--- src/game/g_local.h	(revision 67)
+++ src/game/g_local.h	(working copy)
@@ -1542,6 +1542,15 @@
 extern  vmCvar_t  g_teamStatus;
 extern  vmCvar_t  g_antiSpawnBlock;
 
+extern  vmCvar_t  g_invasion;
+extern  vmCvar_t  g_invasionStartTime;
+extern  vmCvar_t  g_invasionLayout;
+extern  vmCvar_t  g_invasionCurrentLayout;
+extern  vmCvar_t  g_invasionBotNumber;
+extern  vmCvar_t  g_invasionBotMin;
+extern  vmCvar_t  g_invasionBotMax;
+extern  vmCvar_t  g_invasionBotPrefix;
+
 extern  vmCvar_t  g_dretchPunt;
 
 extern  vmCvar_t  g_devmapNoGod;
Index: src/game/g_main.c
===================================================================
--- src/game/g_main.c	(revision 67)
+++ src/game/g_main.c	(working copy)
@@ -23,7 +23,8 @@
 
 #include "g_local.h"
 
-#define QVM_NAME       "Slackers QVM" " (Lakitu7 5.5)"
+#define QVM_NAME       "Slackers QVM" QVM_MODNAME " (Lakitu7 5.5)"
+#define QVM_MODNAME    " (Invasion 2.1)"
 #define QVM_VERSIONNUM      "1.1"
 
 level_locals_t  level;
@@ -198,6 +199,15 @@
 
 vmCvar_t  g_tag;
 
+vmCvar_t  g_invasion;
+vmCvar_t  g_invasionStartTime;
+vmCvar_t  g_invasionLayout;
+vmCvar_t  g_invasionCurrentLayout;
+vmCvar_t  g_invasionBotNumber;
+vmCvar_t  g_invasionBotMin;
+vmCvar_t  g_invasionBotMax;
+vmCvar_t  g_invasionBotPrefix;
+
 vmCvar_t  g_dretchPunt;
 
 vmCvar_t  g_allowShare;
@@ -445,6 +455,15 @@
   
   { &g_tag, "g_tag", "main", CVAR_INIT, 0, qfalse },
   
+  { &g_invasion, "g_invasion", "0", CVAR_ROM, 0, qfalse  },
+  { &g_invasionStartTime, "g_invasionStartTime", "60", CVAR_SERVERINFO | CVAR_ARCHIVE, 0, qfalse  },
+  { &g_invasionLayout, "g_invasionLayout", "invasion", CVAR_ARCHIVE, 0, qfalse  },
+  { &g_invasionCurrentLayout, "g_invasionCurrentLayout", "", CVAR_SERVERINFO | CVAR_ROM, 0, qfalse  },
+  { &g_invasionBotNumber, "g_invasionBotNumber", "", CVAR_ROM, 0, qfalse  },
+  { &g_invasionBotMin, "g_invasionBotMin", "4", CVAR_SERVERINFO | CVAR_ARCHIVE, 0, qfalse  },
+  { &g_invasionBotMax, "g_invasionBotMax", "12", CVAR_SERVERINFO | CVAR_ARCHIVE, 0, qfalse  },
+  { &g_invasionBotPrefix, "g_invasionBotPrefix", "", CVAR_ROM, 0, qfalse  },
+  
   { &g_dretchPunt, "g_dretchPunt", "1", CVAR_ARCHIVE, 0, qfalse  },
   
   { &g_msg, "g_msg", "", CVAR_ARCHIVE, 0, qfalse  },
@@ -983,6 +1002,20 @@
     level.humanTeamLocked=qtrue;
     trap_Cvar_Set( "g_lockTeamsAtStart", "0" );
   }
+
+  //Load invasion OR Reset invasion var
+  if( (level.alienTeamLocked == qtrue) && (level.humanTeamLocked == qtrue) && (g_invasion.integer) &&
+    !Q_strncmp( level.layout, Q_strlwr(g_invasionLayout.string), strlen(g_invasionLayout.string) ) )
+  {
+    level.humanTeamLocked = qfalse;
+    if( g_invasionStartTime.integer > 0 )
+      trap_SendServerCommand( -1, va( "cp \"^1Don't leave your base before %isec !\n\"", g_invasionStartTime.integer ) );
+  }
+  else
+  {
+    trap_Cvar_Set( "g_invasion", 0);
+    trap_Cvar_Set( "g_invasionCurrentLayout" , "" );
+  }
 }
 
 /*
@@ -1373,7 +1406,10 @@
 {
   int i;
   gentity_t *ent;
+  char message[ MAX_STRING_CHARS ];
+  int lastNumAliensSpawns;
 
+  lastNumAliensSpawns = level.numAlienSpawns;
   level.numAlienSpawns = 0;
   level.numHumanSpawns = 0;
 
@@ -1392,6 +1428,20 @@
   //let the client know how many spawns there are
   trap_SetConfigstring( CS_SPAWNS, va( "%d %d",
         level.numAlienSpawns, level.numHumanSpawns ) );
+
+ //display how many eggs there are for the scenario Invasion
+  if( ( lastNumAliensSpawns > level.numAlienSpawns ) &&
+      ( lastNumAliensSpawns >= 0 ) && g_invasion.integer )
+  {
+    if( level.numAlienSpawns == 0 )
+      Com_sprintf( message, sizeof( message ), "^1No egg left... Kill all aliens to finish\n" );
+    else if( level.numAlienSpawns == 1 )
+      Com_sprintf( message, sizeof( message ), "^3Just 1 egg left... you can do it !!!\n" );
+    else
+      Com_sprintf( message, sizeof( message ), "^2%i eggs left\n", level.numAlienSpawns );
+    trap_SendServerCommand( -1, va( "cp \"%s\"", message ) );
+    trap_SendServerCommand( -1, va( "print \"^7CP: %s", message ) );
+  }
 }
 
 /*
@@ -2447,6 +2497,13 @@
     return;
   }
 
+  // if scenario Invasion, go now
+  if( g_invasion.integer )
+  {
+    ExitLevel( );
+    return;
+  }
+
   // if nobody wants to go, clear timer
   if( !ready && numPlayers )
   {
@@ -2513,6 +2570,8 @@
 */
 void CheckExitRules( void )
 {
+  char map[ MAX_QPATH ];
+  trap_Cvar_VariableStringBuffer( "mapname", map, sizeof( map ) );
   // if at the intermission, wait for all non-bots to
   // signal ready, then go to next level
   if( level.intermissiontime )
@@ -2564,7 +2623,12 @@
   {
     //humans win
     level.lastWin = PTE_HUMANS;
-    trap_SendServerCommand( -1, "print \"Humans win\n\"");
+    if (g_invasion.integer)
+      trap_SendServerCommand( -1, va( "print \"\n^3Congratulations !\n^7Humans stopped the "
+        "invasion of ^2%i^7 aliens on the layout ^2%s^7 of the planet ^2%s^7.\n\"",
+        g_invasionBotNumber.integer, g_invasionCurrentLayout.string, map ) );
+    else
+      trap_SendServerCommand( -1, "print \"Humans win\n\"");
     trap_SetConfigstring( CS_WINNER, "Humans Win" );
     LogExit( "Humans win." );
     G_admin_maplog_result( "h" );
@@ -2576,7 +2640,12 @@
   {
     //aliens win
     level.lastWin = PTE_ALIENS;
-    trap_SendServerCommand( -1, "print \"Aliens win\n\"");
+    if (g_invasion.integer)
+      trap_SendServerCommand( -1, va( "print \"\nThe layout ^2%s"
+        "^7 of the planet ^2%s^7 has been destroyed by an invasion of ^2%i^7 aliens.\n\"",
+        g_invasionCurrentLayout.string, map, g_invasionBotNumber.integer) );
+    else
+      trap_SendServerCommand( -1, "print \"Aliens win\n\"");
     trap_SetConfigstring( CS_WINNER, "Aliens Win" );
     LogExit( "Aliens win." );
     G_admin_maplog_result( "a" );
@@ -2949,6 +3018,42 @@
 }
 
 /*
+==================
+CheckInvasionBots
+
+Check if it's time to add Bots for Invasion
+==================
+*/
+void CheckInvasionBots( void )
+{
+  static int invasionBotNumber = 0;
+
+  //wait 5 seconds before adding 1 bot and add 1 bot each 0,5 second
+  if( g_invasion.integer && ( invasionBotNumber < g_invasionBotNumber.integer ) &&
+    ( ( level.time - level.startTime ) >= ( ( 5 * 1000 ) + ( 500 * invasionBotNumber ) ) ) )
+  {
+    char invasionBotName[ MAX_NAME_LENGTH ];
+    invasionBotNumber++;
+    Q_strncpyz( invasionBotName, va( "%s%02i", g_invasionBotPrefix.string, invasionBotNumber ), sizeof( invasionBotName ) );
+    trap_SendConsoleCommand( EXEC_APPEND,va( "!bot add %s aliens 5\n", invasionBotName ) );
+  }
+}
+
+/*
+==================
+CheckInvasionStart
+
+Check if it's time to start Invasion
+==================
+*/
+void CheckInvasionStart( void )
+{
+  if( ( g_invasionStartTime.integer > 0 ) && g_invasion.integer && (level.numHumanSpawns != 0) &&
+    ( ( level.time - level.startTime ) >= ( g_invasionStartTime.integer * 1000 ) ) )
+    trap_SendConsoleCommand( EXEC_APPEND, "!invasion start\n" );
+}
+
+/*
 ================
 G_RunFrame
 
@@ -3116,6 +3221,11 @@
   CheckTeamVote( PTE_HUMANS );
   CheckTeamVote( PTE_ALIENS );
 
+  // check invasion Add Bots
+  CheckInvasionBots( );
+  // check invasion Start
+  CheckInvasionStart( );
+
   // for tracking changes
   CheckCvars( );
 
